### PR TITLE
Resource Kernel

### PR DESCRIPTION
**Pull request by mdavis36**
# Summary

- This PR is a feature
- It does the following :
  - Adds resource parameter to kernel call. Kernel can now be called with `kernel` `kernel_param` `kernel_resource` and `kernel_param_resource`
  - Adds basic kernel loop tests and nested-loop tests with `kernel_resource`
  - Adds nested loop param tests with `kernel_param_resource`
  - All kernel calls return an Event object similar to RAJA::forall

This is most useful with for executing any CUDA/HIP kernel policies on a specific stream.

### CODE REVIEW

**Comment by mdavis36** [comment]
LGTM

**Comment by mdavis36** [comment]
LGTM

**Review by artv3 — DISMISSED** [review]

File: test/functional/kernel/basic-single-loop/CMakeLists.txt [hunk 1]
```diff
@@ -25,6 +25,17 @@ foreach( BACKEND ${KERNEL_BACKENDS} )
     target_include_directories(test-kernel-basic-single-loop-${TESTTYPE}-${BACKEND}.exe
                                PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/tests)
   endforeach()
+
```
Comment (rhornung67) [inline test/functional/kernel/basic-single-loop/CMakeLists.txt:28]: We could consolidate this to cover all tests in one place by adding another nested loop and another variable, right?

set(RESOURCE_USES basic resource-basic)
....

then add innermost loop:

foreach( RESOURCE_USE ${RESOURCE_USES} )
   configure_file( .... test-kernel-${RESOURCE_USE}-single-loop....

File: test/functional/kernel/basic-single-loop/tests/test-kernel-resource-basic-single-loop-segments.hpp [hunk 2]
```diff
@@ -0,0 +1,138 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-21, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
```
Comment (rhornung67) [inline test/functional/kernel/basic-single-loop/tests/test-kernel-resource-basic-single-loop-segments.hpp:7]: How much overlap is there between this file and the non-resource test file? Can that be reduced and invoke the different tests using a simple template parameter type?

File: test/functional/kernel/nested-loop/CMakeLists.txt [hunk 3]
```diff
@@ -20,6 +21,18 @@ foreach( NESTED_LOOP_BACKEND ${KERNEL_BACKENDS} )
     target_include_directories(test-kernel-nested-loop-${NESTED_LOOP_TYPE}-${NESTED_LOOP_BACKEND}.exe
                                PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/tests)
   endforeach()
+
```
Comment (rhornung67) [inline test/functional/kernel/nested-loop/CMakeLists.txt:24]: Similar comment to the one I made on the other CMakelists.txt file.

File: test/functional/kernel/nested-loop/tests/test-kernel-resource-nested-loop-Basic.hpp [hunk 4]
```diff
@@ -0,0 +1,230 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-21, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
```
Comment (rhornung67) [inline test/functional/kernel/nested-loop/tests/test-kernel-resource-nested-loop-Basic.hpp:7]: Same comments/questions as in the other test directory.

**Review by rhornung67 — COMMENTED** [review]
@mdavis36 a few comments/questions to consider to reduce test code redundancy.

File: test/functional/kernel/basic-single-loop/CMakeLists.txt (hunk 1 above)
Comment (mdavis36) [reply test/functional/kernel/basic-single-loop/CMakeLists.txt:28]: Yeah, I went back and forth on the current implementation and what you suggested, I didn't really like either and landed on the more verbose of the two. I can go ahead and use this suggested method if that is preferred

File: test/functional/kernel/basic-single-loop/tests/test-kernel-resource-basic-single-loop-segments.hpp (hunk 2 above)
Comment (mdavis36) [reply test/functional/kernel/basic-single-loop/tests/test-kernel-resource-basic-single-loop-segments.hpp:7]: They're fairly similar, I'll have a go at consolidating the two test implementations into one.

File: test/functional/kernel/basic-single-loop/CMakeLists.txt (hunk 1 above)
Comment (rhornung67) [reply test/functional/kernel/basic-single-loop/CMakeLists.txt:28]: Sounds like you thought about it more than I have. Keep what you have.

File: test/functional/kernel/basic-single-loop/tests/test-kernel-resource-basic-single-loop-segments.hpp (hunk 2 above)
Comment (rhornung67) [reply test/functional/kernel/basic-single-loop/tests/test-kernel-resource-basic-single-loop-segments.hpp:7]: If it starts to get too convoluted, stick with what you have.

Comment (mdavis36) [reply test/functional/kernel/basic-single-loop/tests/test-kernel-resource-basic-single-loop-segments.hpp:7]: Sounds good, I'll let you know

**Comment by mdavis36** [comment]
@rhornung67 I've merged the resource and non resource tests into single implementations, I also ended up changing the cmake files if you'd like to take another look

**Comment by mdavis36** [comment]
LGTM

File: include/RAJA/policy/cuda/kernel/CudaKernel.hpp [hunk 5]
```diff
@@ -466,6 +466,9 @@ struct StatementExecutor<
     using launch_t = CudaLaunchHelper<LaunchConfig, stmt_list_t, data_t, Types>;
 
 
+    RAJA::resources::Cuda res = data.get_resource();
```
Comment (MrBurmark) [inline include/RAJA/policy/cuda/kernel/CudaKernel.hpp:469]: What about Hip?

File: include/RAJA/util/resource.hpp [hunk 6]
```diff
@@ -40,12 +40,26 @@ namespace RAJA
...
   template<typename ExecPol>
-  constexpr auto get_default_resource() -> typename get_resource<ExecPol>::type {
-    return get_resource<ExecPol>::type::get_default();
+  constexpr auto get_default_resource() -> resource_from_pol_t<ExecPol> {
+    return get_resource_from_platform<detail::get_platform<ExecPol>::value>::type::get_default();
   }
 
 #if defined(RAJA_CUDA_ACTIVE)
+  template<>
+  struct get_resource_from_platform<Platform::cuda>{
```
Comment (MrBurmark) [inline include/RAJA/util/resource.hpp:59]: what about hip?

File: include/RAJA/policy/cuda/kernel/CudaKernel.hpp (hunk 5 above)
Comment (mdavis36) [reply include/RAJA/policy/cuda/kernel/CudaKernel.hpp:469]: Added.

File: include/RAJA/util/resource.hpp (hunk 6 above)
Comment (mdavis36) [reply include/RAJA/util/resource.hpp:59]: Added.

**Comment by mdavis36** [comment]
@MrBurmark @rhornung67 I added the changes to work with Hip, everything compiles fine but it seems the Hip nested-loop Multiple Lambda tests are failing, they are also failing on Develop but I'm not sure why yet

**Comment by rhornung67** [comment]
LGTM

**Review by rhornung67 — APPROVED** [review]
@mdavis36 this looks good. This will be merged before my openmp target test subset PR. After this goes in and that branch gets updated, can you please filter out some of these tests similar to what is done for other tests in there? Please let me know if you have any questions.

**Comment by mdavis36** [comment]
LGTM

**Comment by mdavis36** [comment]
LGTM

File: include/RAJA/util/resource.hpp [hunk 7]
```diff
@@ -40,12 +40,26 @@ namespace RAJA
...
+  };
+
+
   template<typename ExecPol>
-  constexpr auto get_default_resource() -> typename get_resource<ExecPol>::type {
-    return get_resource<ExecPol>::type::get_default();
+  using resource_from_pol_t = typename get_resource_from_platform<detail::get_platform<ExecPol>::value>::type;
+
+  template<typename ExecPol>
+  constexpr auto get_default_resource() -> resource_from_pol_t<ExecPol> {
```
Comment (MrBurmark) [inline include/RAJA/util/resource.hpp:53]: Should this return by reference?

**Comment by mdavis36** [comment]
LGTM

**Review by artv3 — DISMISSED** [review]

**Comment by mdavis36** [comment]
LGTM

**Comment by kab163** [comment]
LGTM for the win

**Comment by kab163** [comment]
LGTM

**Comment by mdavis36** [comment]
@rhornung67 This is all green I had some issues with running jobs on gitlab CI and I think it's a permission issue since being added as a member. @kab163 helped get the jobs to run in the meantime. Is this good to go, or do you think we should investigate this gitlab issue further before merging this?

**Review by rhornung67 — APPROVED** [review]
@mdavis36 go ahead and merge this. we'll address gitlab issues if they persist.

**Pull request merged**